src/signature.sh

```diff
@@ -0,0 +1,402 @@
+  else
+    signature="$(str_strip "$signature")"
+    if [[ "$keyword" != 'FIXES' ]] && ! is_valid_signature "$signature"; then
+      return 22 # EINVAL
+    fi
+    trailer+="$signature"
+  fi
+
+  all_trailers+=("$trailer")
+}
+
+# This function receives a commit SHA and verifies if it's
+# a valid commit reference. If it is, then it outputs the
+# appropriate formatted message. Else it returns an error code.
+#
+# @sha Holds either a commit hash or pointer
+#
+# Return:
+# In case of successful return 0 and prints the formatted message,
+# otherwise, return 22.
+function format_fixes_message()
+{
+  local sha="$1"
+  local formatted_message
+
+  # Check if given value is a valid commit reference
+  if [[ $(git cat-file -t "$sha" 2> /dev/null) != 'commit' ]]; then
+    return 22 # EINVAL
+  fi
+
+  # The 'Fixes:' trailer line must follow a format defined by Linux Kernel developers.
+  # Fixes: e21d2170f366 ("video: remove unnecessary platform_set_drvdata()")
+  formatted_message=$(git log -1 "$sha" --oneline --abbrev-commit --abbrev=12 \
+    --format="%h (\\\"%s\\\")")
+
+  printf '%s' "$formatted_message"
+}
+
+# It gets user.name and user.email from git's configuration.
+# If either the name or email are not configured then this
+# function will return an error code. Otherwise it will output
+# the formatted name and email.
+#
+# Return:
+# In case of successful return 0 and prints a properly formatted
+# output, otherwise, return 1.
+function format_name_email_from_user()
+{
+  local user_name
+  local user_email
+
+  user_name="$(git config user.name)"
+  user_email="$(git config user.email)"
+
+  # If user doesn't have either a name or email configured with
+  # git then they must provide an argument
+  if [[ -z "$user_name" || -z "$user_email" ]]; then
+    return 1
+  fi
+
+  printf '%s' "${user_name} <${user_email}>"
+}
+
+# This function writes all the trailer lines stored in @all_trailers
+# into either a patch file or commit. It prints a warning if the
+# given argument is neither a valid patch path nor a commit reference
+# and then ignores it.
+#
+# @patch_or_sha Holds either a patch path or commit SHA
+# @flag Used to specify how 'cmd_manager' will be executed
+function write_all_trailers()
+{
+  local patch_or_sha="$1"
+  local flag="$2"
+  local cmd
+
+  for trailer in "${all_trailers[@]}"; do
+    # Check if given argument is either a patch or valid commit reference,
+    # then build the correct command.
+    if [[ "$(git cat-file -t "$patch_or_sha" 2> /dev/null)" == 'commit' ]]; then
+      cmd="git commit --quiet --amend --no-edit --trailer \"${trailer}\""
+      # Only call 'git rebase' if user is trying to write multiple commits
+      if [[ "$(git rev-parse "$patch_or_sha")" != "$(git rev-parse HEAD)" ]]; then
+        cmd="git rebase ${patch_or_sha} --exec '${cmd}' 2> /dev/null"
+      fi
+    elif is_a_patch "$patch_or_sha"; then
+      cmd="git interpret-trailers ${patch_or_sha} --in-place --trailer \"${trailer}\""
```

Comment:
I didn't know this command before, but I just realized that this is a powerful tool. I thought we could use it with the `--where` parameter to define some heuristics. For example, I really want that at least one Signed-off-by be at the end when we use tags like Tested-by, Reviewed-by, etc. This week, when I used your feature to add the tag `tested-by`, it was added at the end; if it had been added before the last Signed-off-by, it would have been perfect for my use case.

Reply:
The `--where` option completely went under my radar. Looks like a nice tool to use. Unfortunately this seems to be limited to `git interpret-trailers` while `git commit --amend` can't benefit from it. However, I've been working on some heuristics to sort the tags **before** writing them using this command, but as you mentioned all new trailers are written at the end by default and only `interpret-trailers` can change this behavior. I'll be revisiting this topic later because this is a very powerful feature in fact.

I've looked up some solutions to work with commits but all of them comes down to using `sed` in very obscure ways, directly editing the commit messages automatically with `rebase` and `amend`. But I'm too afraid to try this, because I have no idea if I'll be able to come up with a good way to implement that, and also it sounds extremely dangerous.